src/Okta.Sdk/DefaultRetryStrategy.cs

```diff
@@ -0,0 +1,134 @@
+﻿// <copyright file="DefaultRetryStrategy.cs" company="Okta, Inc">
+// Copyright (c) 2014 - present Okta, Inc. All rights reserved.
+// Licensed under the Apache 2.0 license. See the LICENSE file in the project root for full license information.
+// </copyright>
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+
+namespace Okta.Sdk
+{
+    /// <summary>
+    /// Default retry strategy
+    /// </summary>
+    public class DefaultRetryStrategy : IRetryStrategy
+    {
+        /// <summary>
+        /// The default delta used in the back-off formula to account for some clock skew in our service
+        /// </summary>
+        public static readonly int DefaultBackOffSecondsDelta = 1;
+
+        /// <inheritdoc/>
+        public int MaxRetries { get; set; }
+
+        /// <inheritdoc/>
+        public int RequestTimeOut { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DefaultRetryStrategy"/> class.
+        /// </summary>
+        /// <param name="maxRetries">the number of times to retry</param>
+        /// <param name="requestTimeOut">The request timeout in seconds</param>
+        public DefaultRetryStrategy(int maxRetries, int requestTimeOut)
+        {
+            MaxRetries = maxRetries;
+            RequestTimeOut = requestTimeOut;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DefaultRetryStrategy"/> class.
+        /// </summary>
+        public DefaultRetryStrategy()
+        {
+        }
+
+        /// <inheritdoc/>
+        public async Task<HttpResponseMessage> WaitAndRetryAsync(HttpRequestMessage request, Func<HttpRequestMessage, Task<HttpResponseMessage>> operation)
+        {
+            if (request == null)
+            {
+                throw new ArgumentNullException(nameof(request));
+            }
+
+            if (operation == null)
+            {
+                throw new ArgumentNullException(nameof(operation));
+            }
+
+            var attempts = 0;
+            var stopwatch = new Stopwatch();
+            stopwatch.Start();
+
+            do
+            {
+                attempts++;
+
+                var response = await operation(request).ConfigureAwait(false);
+                var statusCode = (int)response.StatusCode;
+
+                try
+                {
+                    if (statusCode == 429 && attempts < MaxRetries + 1 &&
+                        ((RequestTimeOut > 0 && stopwatch.Elapsed.Seconds < RequestTimeOut) || RequestTimeOut == 0))
+                    {
+                        await CreateDelayAsync(response).ConfigureAwait(false);
+                        response.Headers.TryGetValues("X-Okta-Request-Id", out var requestId);
+                        AddRetryOktaHeaders(request, requestId.First(), attempts);
```

Comment:
This call to `requestId.First()` is going to throw if the `X-Okta-Request-Id` header wasn't present in the response. That can be a problem if, say, the request failed due to a network timeout or was rejected by Okta before it actually reached monolith (like if Nginx or an AWS network appliance fails with a 502).

We probably need to change it to `FirstOrDefault()` and handle an empty value in `AddRetryOktaHeaders`.